File: core/registry_save_resolver.py
"""
Resolve PC Gaming Wiki / in-game save locations that point at the Windows registry.

Some games store the save *folder* in a REG_SZ (or only document a registry key); we read
the key and return a normal directory path when it exists on disk.
"""

from __future__ import annotations

import json
import os
import re
import winreg
from pathlib import Path
from typing import Dict, List, Optional, Tuple

from core.save_location_fetcher import path_to_directory_only

# Built-in fallbacks when PCGW has no data (extend via config/steam_registry_save_keys.json).
STEAM_APP_REGISTRY_SAVE_KEYS: Dict[str, Tuple[str, str]] = {
    "1843760": ("HKEY_CURRENT_USER", r"Software\Die of Death Games\Rogue Tower"),
}

_STEAM_REGISTRY_MERGED_CACHE: Optional[Dict[str, Tuple[str, str]]] = None


def merged_steam_registry_save_keys() -> Dict[str, Tuple[str, str]]:
    """Defaults plus optional JSON map: app id string -> (hive_name, subkey)."""
    global _STEAM_REGISTRY_MERGED_CACHE
    if _STEAM_REGISTRY_MERGED_CACHE is not None:
        return _STEAM_REGISTRY_MERGED_CACHE
    out: Dict[str, Tuple[str, str]] = dict(STEAM_APP_REGISTRY_SAVE_KEYS)
    cfg = Path(__file__).resolve().parent.parent / "config" / "steam_registry_save_keys.json"
    if cfg.is_file():
        try:
            raw = json.loads(cfg.read_text(encoding="utf-8"))
            if isinstance(raw, dict):
                for k, v in raw.items():
                    aid = str(k).strip()
                    if not aid or not isinstance(v, dict):
                        continue
                    hive = str(v.get("hive", "")).strip()
                    sub = str(v.get("subkey", "")).strip()
                    if hive and sub:
                        out[aid] = (hive, sub.replace("/", "\\"))
        except (json.JSONDecodeError, OSError):
            pass
    _STEAM_REGISTRY_MERGED_CACHE = out
    return out

_HIVE_PREFIXES: tuple[tuple[str, int], ...] = (
    ("HKEY_CURRENT_USER\\", winreg.HKEY_CURRENT_USER),
    ("HKCU\\", winreg.HKEY_CURRENT_USER),
    ("HKEY_LOCAL_MACHINE\\", winreg.HKEY_LOCAL_MACHINE),
    ("HKLM\\", winreg.HKEY_LOCAL_MACHINE),
    ("HKEY_USERS\\", winreg.HKEY_USERS),
    ("HKU\\", winreg.HKEY_USERS),
    ("HKEY_CLASSES_ROOT\\", winreg.HKEY_CLASSES_ROOT),
    ("HKCR\\", winreg.HKEY_CLASSES_ROOT),
)

# Order: try explicit value name(s), then default.
_VALUE_NAMES_TO_TRY: tuple[str, ...] = (
    "SavePath",
    "Savegame",
    "SaveDirectory",
    "Save Dir",
    "SaveDir",
    "SaveGame",
    "PlayerSaveDir",
    "InstallPath",
    "Path",
    "SaveLocation",
    "DataPath",
    "",
)

_REGISTRY_LINE_RE = re.compile(
    r"(?is)\b("
    r"HKEY_(?:CURRENT_USER|LOCAL_MACHINE|USERS|CLASSES_ROOT|PERFORMANCE_DATA)\s*[\\/][^;\n\r|]+"
    r"|(?:HKCU|HKLM|HKCR|HKU)\s*[\\/][^;\n\r|]+"
    r")"
)


def wiki_cell_looks_like_file_path(text: str) -> bool:
    """True if the wiki cell text looks like a filesystem path (not registry-only)."""
    if not text or not isinstance(text, str):
        return False
    t = text.strip()
    if not t:
        return False
    if "%" in t or t.startswith("~/") or "/Users/" in t or "/Library/" in t:
        return True
    if re.search(r"[A-Za-z]:\\", t) or re.search(r"[A-Za-z]:/", t):
        return True
    return False


def looks_like_registry_hive_line(text: str) -> bool:
    if not text or not isinstance(text, str):
        return False
    s = normalize_registry_pasted_path(text)
    if not s:
        return False
    up = s.upper()
    if up.startswith("HKEY_") or up.startswith("HKCU") or up.startswith("HKLM") or up.startswith("HKCR") or up.startswith("HKU"):
        return True
    return bool(_REGISTRY_LINE_RE.search(s))


def normalize_registry_pasted_path(text: str) -> str:
    """Strip ``Computer\\`` prefix from Registry Editor copy-paste."""
    return re.sub(r"(?i)^Computer\\+", "", (text or "").strip())


def extract_registry_hints(text: Optional[str]) -> List[str]:
    """Return distinct registry path substrings found in a wiki cell or paragraph."""
    if not text or not isinstance(text, str):
        return []
    text = normalize_registry_pasted_path(text)
    out: List[str] = []
    seen: set[str] = set()
    for m in _REGISTRY_LINE_RE.finditer(text):
        h = m.group(1).strip().strip('"').replace("/", "\\")
        if h and h not in seen:
            seen.add(h)
            out.append(h)
    return out


def _split_hive_and_remainder(hint: str) -> tuple[Optional[int], str]:
    hint = hint.strip().strip('"').replace("/", "\\")
    for prefix, hive in _HIVE_PREFIXES:
        if hint.upper().startswith(prefix.upper()):
            rest = hint[len(prefix) :].lstrip("\\")
            return hive, rest
    return None, ""


def _value_to_folder(val: object, typ: int) -> Optional[str]:
    if typ == winreg.REG_MULTI_SZ:
        if not isinstance(val, list):
            return None
        parts = [str(x).strip() for x in val if str(x).strip()]
        if not parts:
            return None
        return _value_to_folder(parts[0], winreg.REG_SZ)
    if typ not in (winreg.REG_SZ, winreg.REG_EXPAND_SZ):
        return None
    s = os.path.expandvars(str(val).strip().strip('"'))
    if not s:
        return None
    if os.path.isfile(s):
        s = os.path.dirname(s)
    s = os.path.normpath(s)
    if os.path.isdir(s):
        return path_to_directory_only(s + "\\") if not s.endswith("\\") else path_to_directory_only(s)
    return None


def _try_read_values_from_open_key(key) -> Optional[str]:
    for vn in _VALUE_NAMES_TO_TRY:
        try:
            val, typ = winreg.QueryValueEx(key, vn)
        except OSError:
            continue
        folder = _value_to_folder(val, typ)
        if folder:
            return folder
    return None


def format_registry_save_display(hive_name: str, subkey: str) -> str:
    return f"{hive_name}\\{subkey}"


def try_registry_key_as_in_key_save_location(hint: str) -> Optional[Tuple[str, str]]:
    """
    If ``hint`` opens as a registry *key* whose values hold save state (no folder path),
    return ``(hive_name, subkey)`` for display and ``reg export`` backup.
    """
    hint = normalize_registry_pasted_path(hint)
    hive, remainder = _split_hive_and_remainder(hint)
    if hive is None or not remainder:
        return None
    hive_name = _hive_id_to_name(hive)
    try:
        k = winreg.OpenKey(hive, remainder, 0, winreg.KEY_READ)
    except OSError:
        return None
    try:
        n_values, n_subkeys, _ = winreg.QueryInfoKey(k)
        if n_values < 1 and n_subkeys < 1:
            return None
        return (hive_name, remainder)
    finally:
        try:
            k.Close()
        except Exception:
            pass


def _hive_id_to_name(hk: int) -> str:
    mapping = {
        winreg.HKEY_CURRENT_USER: "HKEY_CURRENT_USER",
        winreg.HKEY_LOCAL_MACHINE: "HKEY_LOCAL_MACHINE",
        winreg.HKEY_USERS: "HKEY_USERS",
        winreg.HKEY_CLASSES_ROOT: "HKEY_CLASSES_ROOT",
    }
    return mapping.get(hk, "HKEY_UNKNOWN")


def resolve_registry_hint_to_save_folder(hint: str) -> Optional[str]:
    """
    Given a line like ``HKEY_CURRENT_USER\\Software\\Vendor\\Game`` or a path that ends
    with a value name under an existing key, return a save *directory* on disk if found.
    """
    hint = normalize_registry_pasted_path(hint)
    hive, remainder = _split_hive_and_remainder(hint)
    if hive is None or not remainder:
        return None
    parts = [p for p in remainder.split("\\") if p]
    if not parts:
        return None

    for depth in range(len(parts), 0, -1):
        key_path = "\\".join(parts[:depth])
        value_segments = parts[depth:]
        try:
            k = winreg.OpenKey(hive, key_path, 0, winreg.KEY_READ)
        except OSError:
            continue
        try:
            if value_segments:
                vn = value_segments[0]
                try:
                    val, typ = winreg.QueryValueEx(k, vn)
                except OSError:
                    pass
                else:
                    folder = _value_to_folder(val, typ)
                    if folder:
                        return folder
            folder = _try_read_values_from_open_key(k)
            if folder:
                return folder
        finally:
            try:
                k.Close()
            except Exception:
                pass
    return None

File: src-winui/src/GSBT.Core/Services/RegistrySaveResolver.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Runtime.Versioning;
using GSBT.Core.Common;
using Microsoft.Win32;

namespace GSBT.Core.Services;

[SupportedOSPlatform("windows")]
public sealed class RegistrySaveResolver
{
    private static readonly Regex RegistryLine = new(
        @"(?is)\b(HKEY_(?:CURRENT_USER|LOCAL_MACHINE|USERS|CLASSES_ROOT)\s*[\\/][^;\n\r|]+|(?:HKCU|HKLM|HKCR|HKU)\s*[\\/][^;\n\r|]+)",
        RegexOptions.Compiled);

    private static readonly string[] ValueNamesToTry =
    [
        "SavePath", "Savegame", "SaveDirectory", "Save Dir", "SaveDir", "SaveGame", "PlayerSaveDir", "InstallPath", "Path", "SaveLocation", "DataPath", ""
    ];

    private readonly Lazy<Dictionary<string, (string Hive, string Subkey)>> _mergedSteamRegistrySaveKeys;

    public RegistrySaveResolver(string? configRoot = null)
    {
        _mergedSteamRegistrySaveKeys = new(() => LoadMergedSteamRegistrySaveKeys(configRoot));
    }

    public IReadOnlyDictionary<string, (string Hive, string Subkey)> MergedSteamRegistrySaveKeys => _mergedSteamRegistrySaveKeys.Value;

    public static string NormalizeRegistryPastedPath(string text)
        => Regex.Replace((text ?? string.Empty).Trim(), @"(?i)^Computer\\+", string.Empty);

    public static bool LooksLikeRegistryHiveLine(string text)
    {
        var s = NormalizeRegistryPastedPath(text);
        if (string.IsNullOrWhiteSpace(s))
        {
            return false;
        }

        var up = s.ToUpperInvariant();
        return up.StartsWith("HKEY_") || up.StartsWith("HKCU") || up.StartsWith("HKLM") || up.StartsWith("HKCR") || up.StartsWith("HKU") || RegistryLine.IsMatch(s);
    }

    /// <summary>Detects pasted folder paths mistaken for registry paths (e.g. <c>C:\Games\…</c>).</summary>
    public static bool LooksLikeFilesystemPath(string text)
    {
        var t = (text ?? string.Empty).Trim().Trim('"');
        if (t.Length == 0)
        {
            return false;
        }

        if (t.StartsWith("file:", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (t.Length >= 2 && char.IsLetter(t[0]) && t[1] == ':')
        {
            return true;
        }

        if (t.StartsWith(@"\\", StringComparison.Ordinal) || t.StartsWith("//", StringComparison.Ordinal))
        {
            return true;
        }

        if (t.StartsWith('\\') && !LooksLikeRegistryHiveLine(t))
        {
            return true;
        }

        return false;
    }

    public enum RegistrySaveValidationKind
    {
        ValidFolder,
        ValidInKey,
        FilesystemPath,
        NotRegistrySyntax,
        UnknownHive,
        MissingSubkeyPath,
        KeyNotFound,
        KeyEmpty,
        FolderTargetMissing,
    }

    public readonly record struct RegistrySaveValidation(
        RegistrySaveValidationKind Kind,
        string Message,
        string? FolderRaw = null,
        string? FolderResolved = null,
        string? Hive = null,
        string? Subkey = null)
    {
        public bool IsSuccess =>
            Kind is RegistrySaveValidationKind.ValidFolder or RegistrySaveValidationKind.ValidInKey;
    }

    /// <summary>Validates a single registry hint before assigning a save location (folder or in-key).</summary>
    public RegistrySaveValidation ValidateRegistrySaveHint(string hint)
    {
        var normalized = NormalizeRegistryPastedPath(hint);
        if (string.IsNullOrWhiteSpace(normalized))
        {
            return new(RegistrySaveValidationKind.NotRegistrySyntax, "Enter a registry path or paste from Regedit.");
        }

        if (LooksLikeFilesystemPath(normalized))
        {
            return new(
                RegistrySaveValidationKind.FilesystemPath,
                "That looks like a file or folder path, not a registry key. Copy the path from Regedit’s address bar (e.g. HKCU\\Software\\…).");
        }

        if (!LooksLikeRegistryHiveLine(normalized))
        {
            return new(
                RegistrySaveValidationKind.NotRegistrySyntax,
                "Could not parse a registry path. Example: HKCU\\Software\\YourGame");
        }

        var (hive, remainder) = SplitHiveAndRemainder(normalized);
        if (hive is null)
        {
            if (IsKnownHiveTokenOnly(normalized))
            {
                return new(
                    RegistrySaveValidationKind.MissingSubkeyPath,
                    "Include the full key path after the hive (e.g. HKCU\\Software\\YourGame).");
            }

            return new(
                RegistrySaveValidationKind.UnknownHive,
                "Unrecognized registry hive. Use HKCU, HKLM, HKU, or HKCR.");
        }

        if (string.IsNullOrWhiteSpace(remainder))
        {
            return new(
                RegistrySaveValidationKind.MissingSubkeyPath,
                "Include the full key path after the hive (e.g. HKCU\\Software\\YourGame).");
        }

        var folder = ResolveRegistryHintToSaveFolder(hint);
        if (!string.IsNullOrWhiteSpace(folder))
        {
            try
            {
                var expanded = Environment.ExpandEnvironmentVariables(folder.Trim());
                var resolved = Path.GetFullPath(expanded);
                if (!Directory.Exists(resolved))
                {
                    return new(
                        RegistrySaveValidationKind.FolderTargetMissing,
                        $"Registry points to a folder that does not exist:\n{resolved}",
                        FolderRaw: folder.Trim(),
                        FolderResolved: resolved);
                }

                return new(
                    RegistrySaveValidationKind.ValidFolder,
                    string.Empty,
                    FolderRaw: folder.Trim(),
                    FolderResolved: resolved);
            }
            catch (Exception ex)
            {
                return new(
                    RegistrySaveValidationKind.FolderTargetMissing,
                    $"Registry points to an invalid folder path: {ex.Message}",
                    FolderRaw: folder.Trim());
            }
        }

        using var key = hive.OpenSubKey(remainder, writable: false);
        if (key is null)
        {
            return new(
                RegistrySaveValidationKind.KeyNotFound,
                "Registry key not found. In Regedit, select the key (folder icon), not a single value, and copy the path from the address bar.");
        }

        if (key.ValueCount < 1 && key.SubKeyCount < 1)
        {
            return new(
                RegistrySaveValidationKind.KeyEmpty,
                "That registry key exists but has no values or subkeys to back up.");
        }

        return new(
            RegistrySaveValidationKind.ValidInKey,
            string.Empty,
            Hive: HiveName(hive),
            Subkey: remainder);
    }

    public static IReadOnlyList<string> ExtractRegistryHints(string? text)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return [];
        }

        var cleaned = NormalizeRegistryPastedPath(text);
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var list = new List<string>();
        foreach (Match m in RegistryLine.Matches(cleaned))
        {
            var hint = m.Groups[1].Value.Trim().Trim('"').Replace('/', '\\');
            if (seen.Add(hint))
            {
                list.Add(hint);
            }
        }

        return list;
    }

    public static string FormatRegistrySaveDisplay(string hive, string subkey) => $"{hive}\\{subkey}";

    public string? ResolveRegistryHintToSaveFolder(string hint)
    {
        var normalized = NormalizeRegistryPastedPath(hint);
        var (hive, remainder) = SplitHiveAndRemainder(normalized);
        if (hive is null || string.IsNullOrWhiteSpace(remainder))
        {
            return null;
        }

        var parts = remainder.Split('\\', StringSplitOptions.RemoveEmptyEntries);
        for (var depth = parts.Length; depth >= 1; depth--)
        {
            var keyPath = string.Join('\\', parts.Take(depth));
            var valueSegments = parts.Skip(depth).ToArray();
            using var key = hive.OpenSubKey(keyPath, writable: false);
            if (key is null)
            {
                continue;
            }

            if (valueSegments.Length > 0)
            {
                var fromValue = ValueToFolder(key.GetValue(valueSegments[0]));
                if (!string.IsNullOrWhiteSpace(fromValue))
                {
                    return fromValue;
                }
            }

            foreach (var vn in ValueNamesToTry)
            {
                var folder = ValueToFolder(key.GetValue(vn));
                if (!string.IsNullOrWhiteSpace(folder))
                {
                    return folder;
                }
            }
        }

        return null;
    }

    public (string Hive, string Subkey)? TryRegistryKeyAsInKeySaveLocation(string hint)
    {
        var normalized = NormalizeRegistryPastedPath(hint);
        var (hive, remainder) = SplitHiveAndRemainder(normalized);
        if (hive is null || string.IsNullOrWhiteSpace(remainder))
        {
            return null;
        }

        using var key = hive.OpenSubKey(remainder, writable: false);
        if (key is null)
        {
            return null;
        }

        if (key.ValueCount < 1 && key.SubKeyCount < 1)
        {
            return null;
        }

        return (HiveName(hive), remainder);
    }

    private static Dictionary<string, (string Hive, string Subkey)> LoadMergedSteamRegistrySaveKeys(string? configRoot)
    {
        var outMap = new Dictionary<string, (string Hive, string Subkey)>(StringComparer.OrdinalIgnoreCase);

        var root = configRoot ?? Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", ".."));
        var jsonPath = Path.Combine(root, "config", "steam_registry_save_keys.json");
        if (!File.Exists(jsonPath))
        {
            return outMap;
        }

        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(jsonPath));
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                return outMap;
            }

            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                if (prop.Value.ValueKind != JsonValueKind.Object)
                {
                    continue;
                }

                var hive = prop.Value.TryGetProperty("hive", out var h) ? h.GetString() : null;
                var sub = prop.Value.TryGetProperty("subkey", out var s) ? s.GetString() : null;
                if (string.IsNullOrWhiteSpace(hive) || string.IsNullOrWhiteSpace(sub))
                {
                    continue;
                }

                outMap[prop.Name] = (hive.Trim(), sub.Replace('/', '\\'));
            }
        }
        catch
        {
            // Best effort parse.
        }

        return outMap;
    }

    private static string? ValueToFolder(object? value)
    {
        if (value is null)
        {
            return null;
        }

        if (value is string[] arr && arr.Length > 0)
        {
            return ValueToFolder(arr[0]);
        }

        if (value is not string raw)
        {
            return null;
        }

        var s = Environment.ExpandEnvironmentVariables(raw.Trim().Trim('"'));
        if (string.IsNullOrWhiteSpace(s))
        {
            return null;
        }

        if (File.Exists(s))
        {
            s = Path.GetDirectoryName(s) ?? s;
        }

        try
        {
            s = Path.GetFullPath(s);
        }
        catch
        {
            return null;
        }

        if (!Directory.Exists(s))
        {
            return null;
        }

        var candidate = s.EndsWith('\\') ? s : $"{s}\\";
        return PathUtils.PathToDirectoryOnly(candidate);
    }

    private static string HiveName(RegistryKey key)
    {
        if (ReferenceEquals(key, Registry.CurrentUser))
        {
            return "HKEY_CURRENT_USER";
        }

        if (ReferenceEquals(key, Registry.LocalMachine))
        {
            return "HKEY_LOCAL_MACHINE";
        }

        if (ReferenceEquals(key, Registry.Users))
        {
            return "HKEY_USERS";
        }

        if (ReferenceEquals(key, Registry.ClassesRoot))
        {
            return "HKEY_CLASSES_ROOT";
        }

        return "HKEY_UNKNOWN";
    }

    private static bool IsKnownHiveTokenOnly(string normalized)
    {
        var token = normalized.Trim().TrimEnd('\\');
        return token.ToUpperInvariant() switch
        {
            "HKCU" or "HKLM" or "HKCR" or "HKU" => true,
            "HKEY_CURRENT_USER" or "HKEY_LOCAL_MACHINE" or "HKEY_USERS" or "HKEY_CLASSES_ROOT" => true,
            _ => false,
        };
    }

    private static (RegistryKey? Hive, string Remainder) SplitHiveAndRemainder(string hint)
    {
        var normalized = hint.Trim().Trim('"').Replace('/', '\\');
        var mappings = new (string Prefix, RegistryKey Hive)[]
        {
            ("HKEY_CURRENT_USER\\", Registry.CurrentUser),
            ("HKCU\\", Registry.CurrentUser),
            ("HKEY_LOCAL_MACHINE\\", Registry.LocalMachine),
            ("HKLM\\", Registry.LocalMachine),
            ("HKEY_USERS\\", Registry.Users),
            ("HKU\\", Registry.Users),
            ("HKEY_CLASSES_ROOT\\", Registry.ClassesRoot),
            ("HKCR\\", Registry.ClassesRoot)
        };

        foreach (var (prefix, hive) in mappings)
        {
            if (normalized.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
            {
                return (hive, normalized[prefix.Length..].TrimStart('\\'));
            }
        }

        return (null, string.Empty);
    }
}
